### PR TITLE
Add estimate functionality to handleCheck method.

### PR DESCRIPTION
**Pull request by veeteeran**
## Description
This PR updates the database with the `estimatedNextPurchase` date and `numberOfPurchases.` The function [/src/lib/estimates.js](https://github.com/the-collab-lab/tcl-17-smart-shopping-list/blob/main/src/lib/estimates.js)  is used to generate the number of days until next purchase.

## Related Issue
Closes [#11](https://github.com/the-collab-lab/tcl-17-smart-shopping-list/issues/11)

## Acceptance Criteria
- [x] When a purchase is recorded, the estimated number of days until the next purchase date should be calculated and recorded in the database

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
| ✓ | :sparkles: New feature     |

## Updates
No UI updates for this feature

## Testing Steps / QA Criteria

* One way to test the site is to click on the `details` button below to open the Netlify preview

Screen Shot 2021-01-12 at 6 05 51 PM

* Otherwise from your terminal, pull down this branch with `git pull origin BL-VT-estimate-next-purchase` and check that branch out with `git checkout BL-VT-estimate-next-purchase`
* For PR review `npm install --save date-fns`, after review finished `npm uninstall date-fns`
* `npm start` to launch the app.
* Create a new list and add items or if a list exists click the List button
* Tap a check box of any item on the list
* Open Firebase and confirm the fields `estimatedNextPurchase` and `numberOfPurchases` are updated

### CODE REVIEW

File: src/components/List.js
Comment (LINDGRENBA) [inline src/components/List.js:63]: This code is specifically for our demo on Sunday. We will install the date-fns dependency to demo the times in the console. However, this is not necessary for the functionality and will be removed before merging.

File: src/components/List.js
Comment (LINDGRENBA) [inline src/components/List.js:6]: These date-fns import statements are for the demo only and will be removed before merging into main.

**Review by LINDGRENBA — APPROVED** [review]

**Comment by LINDGRENBA** [comment]
One question @veeteeran and I had is that we are reusing the code `targetDoc[0]` several times. Should we refactor to store this in a variable and use the variable in its place?

**Comment by laurenmbeatty** [comment]
> One question @veeteeran and I had is that we are reusing the code `targetDoc[0]` several times. Should we refactor to store this in a variable and use the variable in its place?

I vote yes.

**Review by rachelrohrbach — APPROVED** [review]
Looks good, way to go on giving our app the "smart" shopping list functionality! I tried the Netlify demo and it seems to be working well and updating the firestore collection accordingly. I appreciated the comments in the code!  
And I vote yes as well to storing targetDoc[0] as a variable since you're reusing it several times in the code.

**Review by DeeDowns — APPROVED** [review]
I also vote for creating a variable for `targetDoc[0]`
Great job figuring this issue, testing it out and it worked perfectly. This issue seemed challenging but you guys did well 👌

File: src/components/List.js
Comment (veeteeran) [inline src/components/List.js:97]: @LINDGRENBA and I think `checked` and `disabled` follow the same logic so passing the `checkedPurchaseDate` function to `disabled` is the route we took. Once checked the checkbox is disabled so a user can't continually tap the checkbox and update the `lastPurchased` date

File: src/components/List.js
Comment (laurenmbeatty) [inline src/components/List.js:36]: I'm wondering, could you just skip the call the db, and just use the `event.target.id` in place of `queryCollection.id`?  Little less costly, I think.

Comment (laurenmbeatty) [reply src/components/List.js:36]: and then below at the `queryCollection.update` part...you could just do:

`db.collection(token).doc(event.target.id).update.....`

File: src/components/List.js
Comment (laurenmbeatty) [inline src/components/List.js:73]: This is a super-tricky one, and I'm trying to understand it, but right now, the estimatedNextPurchase is being saved as milliseconds, correct? Right now this is one of the values that I saved: 1612403611806, which converts to around 18,000 days or 01/13/53065. 

image

I wonder if the three params for the estimates function above need to be 

1. `estimatedNextPurchase || timeFrame` in _days_ (use timeFrame if there hasn't been a calculated value yet
2. latest interval should be the distance between last purchase and this purchase in _days_
3. Number of purchases
Then it should return a number in _days_

I 100% could be wrong.

File: src/components/List.js
Comment (laurenmbeatty) [inline src/components/List.js:66]: Ooo..cool.

File: src/components/List.js
Comment (laurenmbeatty) [reply src/components/List.js:73]: Hmm...If I leave off the last three digits...then it's OK? You know what would be awesome...could you display the estimated value in the # of days next to the item name? Then it would be easier to understand and your reviewer would be super grateful : ). 👍

File: src/components/List.js
Comment (MikeBLambert) [inline src/components/List.js:43]: In functional programming, which is the paradigm that React more or less ascribes to, it is thought of as best practice to avoid mutating data as it can cause non-deterministic behavior really hard to catch bugs. 

In this case, we are mutating `listItem.numberOfPurchases` with `+=`. I think that we can just replace `+=` with `+` as we're really just looking to assign the value of `listItem.numberOfPurchases + 1` to `numberOfPurchases`

Here's a good little article about immutability in functional programming - https://dev.to/mpodlasin/functional-programming-in-js-part-ii-immutability-vanilla-js-immutable-js-and-immer-2ccm

File: src/components/List.js
Comment (MikeBLambert) [reply src/components/List.js:66]: Nice! Love the idea of having something like this ready to help out with the demo. 
One thought that I had is that you could create a demo branch where you would keep this code, then you could merge the intended code changes before the demo without merging this demo code and also have something socked away that you can use for demo purposes.

File: src/components/List.js
Comment (LINDGRENBA) [reply src/components/List.js:73]: That's a great idea, and maybe we could add a screenshot of some of the values we're checking with those comments, too, so you can see them before our demo.

File: src/components/List.js
Comment (LINDGRENBA) [reply src/components/List.js:43]: Yes! Of course 🤦 that's React's whole thing... I think it totally escaped us that that's what we were doing there. Thank you for pointing it out.

File: src/components/List.js
Comment (LINDGRENBA) [reply src/components/List.js:96]: We added functionality to display the duration of time until the next purchase, but had some issues trying to display the actual date of the next purchase. After spending a reasonable amount of time trying to get it to get it to work we decided to move on. 
timeTillNextPurchase-consoleLogs

File: src/components/List.js
Comment (LINDGRENBA) [reply src/components/List.js]: When we change `queryCollection.id` to `event.target.id`  or just try to console.log the value of `event.target.id`, we get the following error :
target-event-id-error

**Comment by chrisdopuch** [comment]
Hey @veeteeran @LINDGRENBA I tried QA testing the app on the netlify demo, but I ran into an issue. After adding an item to my list, I am unable to check the checkbox. Popping open the terminal, I see the following error show up when I try to check the box:

File: src/components/List.js
Comment (chrisdopuch) [inline src/components/List.js:44]: Way to check for the `numberOfPurchases` being undefined 👍 

In many programming languages, there's a thing called a "guard" statement that allows you to express this kind of logic in a more compact way. JS has this feature through the use of our boolean operators (e.g. `||` or `&&`).

You can use those operators to check for certain values truthiness just like we're doing here, but with a shorter syntax.
That would look like:
```
    const numberOfPurchases = (listItem.numberOfPurchases || 0) + 1;
```

This is just a code style kind of thing, so definitely optional. I just wanted to make you aware of it, if you didn't already know! Here's a blog post that goes into more detail if you're curious how it works: https://seanmonstar.com/post/707078771/guard-and-default-operators

File: src/components/List.js
Comment (chrisdopuch) [inline src/components/List.js:124]: It appears this value is missing when I QA tested on netlify:

**Review by chrisdopuch — COMMENTED** [review]
Nice work so far y'all, getting close! 👏 
I'll come back later and approve once the demo is working, just ping me once you've fixed the issue @veeteeran @LINDGRENBA 😄

File: src/components/List.js
Comment (veeteeran) [reply src/components/List.js:44]: I did not know about this. Thanks!

File: src/components/List.js
Comment (veeteeran) [reply src/components/List.js]: @chrisdopuch The above issue has been fixed. Thanks for hopping on Zoom with me earlier. Things seem to be working 🤞🏽 😅

Comment (LINDGRENBA) [reply src/components/List.js]: Thanks for taking care of that! Also, @chrisdopuch , those are some great list items.

Comment (veeteeran) [reply src/components/List.js]: Everyone needs 🧦 🧦  and gummie bears!

**Comment by LINDGRENBA** [comment]
@chrisdopuch it looks like the Netlify issue is resolved, at least when I try it on my end. Let us know if it's still not working for you!

File: src/components/List.js
Comment (chrisdopuch) [inline src/components/List.js:136]: Nice, this is a great UX improvement! 👏

File: src/components/List.js
Comment (chrisdopuch) [inline src/components/List.js:134]: @veeteeran @LINDGRENBA the first time I check the checkbox, the "time until next purchase" text shows up but it doesn't display a value. Should it be displaying something?
purchase

Once I click it a second time, then it starts to show the value

**Review by chrisdopuch — APPROVED** [review]
This is looking good you two! 😄 🎉 
I had one small question about if the value should be shown for the first click of the checkbox
But otherwise, it's working great now!

**Comment by veeteeran** [comment]
@chrisdopuch I realized the condition is written to check if numberOfPurchases is greater than 0. We can change it to greater than 1 and display a message saying "More info need to calculate next purchase"

or something less wordy? 😄

File: src/components/List.js
Comment (LINDGRENBA) [reply src/components/List.js:134]: Hi @chrisdopuch . One thing I wasn't sure was clear is that the value is only there to make it easier for reviewers to see that it's working and won't be part of the final UI/UX. Lauren suggested it in one of the previous comments. I still think this was a good thing to point out because it may be an indicator of the code not functioning as we expected, but it won't be part of the final UI.... unless the team likes it and wants to keep it in, but I think we deal with organizing the items by this value in one of next week's issues and there's a different approach 👍 😄

Comment (MikeBLambert) [reply src/components/List.js:134]: I may be totally misunderstanding things (as I often do... ;-P), but it looks like right now we're saving to the DB a human readable form of the time to next purchase at the exact time that the user checked the checkbox. That means that that value will never change, even as time marches on. I believe that we may only need to save the estimated next purchase date and the number of times that the item has been purchased in the database. 

We should be able to derive the time until the next purchase from the distance between now and the estimated next purchase date and we can just calculate that on the fly via a function in our component. My general rule for determining whether or not something should be saved to a DB (with every rule there are always exceptions...) is that if I can derive it from other pre-existing values, then there's no need to save it. 

However, as you mentioned, this is just for demo purposes, so it's not a biggie either way. :-)

File: src/components/List.js
Comment (laurenmbeatty) [reply src/components/List.js]: Hmmm....I guess when I did it when playing around on your branch I grabbed the value of `event.target.id` right at the beginning of the function:

`const id = event.target.id`

**Comment by laurenmbeatty** [comment]
When I look in Firestore I see:
image

The AC says: "When a purchase is recorded, the _estimated number of days_ until the next purchase date should be calculated and recorded in the database". 

Which of the fields in the db refers to the estimated number of days? I'm expecting to see a whole number value like: 16, 30, 3, 14, --- a number similar to the `timeFrame` field, but maybe the plan is to use one of those numbers and convert it into a value that reflects number of days.

File: src/components/List.js
Comment (MikeBLambert) [reply src/components/List.js:134]: Correcting myself slightly as @laurenmbeatty rightly pointed out we are supposed to be recording the estimated days until next purchase, not the next purchase date. I would take that to mean the number of days, which, if added onto the last purchase date, will give us our next purchase date. Thus, if we want to display how much time a user has at any given moment until they should be purchasing again, we would not want to directly display the value in the DB, as that won't change, but rather we would display the time between now and that next purchase date.

**Comment by laurenmbeatty** [comment]
I've created a sandbox of the problem here, and the kind of values I would expect to pass to the function, and the kind of values I would expect to be returned and saved in firestore:
https://repl.it/join/lptucqyb-laurenbeatty

If you click "Run" it will give you the logged values in the shell: 
image

**Comment by veeteeran** [comment]
Hi @laurenmbeatty. Thanks for pointing us in the right direction. We completely overlooked inputting values as whole numbers and we can definitely rename our variable to be more concise. I've found a function in date-fns `differenceInDays` that should give us the whole number of days to pass as the second argument.
`const latestInterval = differenceInDays(currentDate, lastPurchased);`


> // example if it was the first purchase, so there really wasn't a latest interval
> // I'm passing in whatever the timeFrame was to latestInterval value instead, since getting the latestInterval between the current purchase date and null seems weird

I see what you mean here. we should be checking if `numberOfPurchases < 2` set the latestInterval to the timeFrame else set latestInterval to the whole number of days as figured above

> // What if you're purchasing it on the same day??? what does that give you?
For this question, your testing shows giving 0 as the second argument works. This is great since `differenceInDays` returns 0 if the difference is under 24 hours. I don't think this would be an issue if zero didn't work since we disable the checkbox. At the moment we have that line commented out for the purpose of reviewing.

Sorry, that was a long reply. I hope it wasn't too confusing. 😅  I'll push code as soon as I can

File: src/components/List.js
Comment (veeteeran) [reply src/components/List.js:133]: @MikeBLambert I agree with what @laurenmbeatty and you are saying about the values saved in the database. In the database we want to save `numberOfPurchases` and `daysUntilNextPurchase`.

What is time? 🤣

File: src/components/List.js
Comment (laurenmbeatty) [inline src/components/List.js:42]: Nice!

File: src/components/List.js
Comment (laurenmbeatty) [inline src/components/List.js:47]: This is my only question. My thought for this calculation is that it only uses the `timeFrame` value on the very first purchase, and then subsequently uses the `daysUntilNextPurchase` value. I think the function expects "The last estimated purchase interval", so it makes sense to me to do that because the calculation gets more and more refined as you purchase that item more and more.  What do you all think?

**Review by laurenmbeatty — APPROVED** [review]
I just had one more question, but GOOD JOBBBBBB!!!!!!!

File: src/components/List.js
Comment (veeteeran) [reply src/components/List.js:47]: @laurenmbeatty 💯  We'll change it so the first purchase uses `timeFrame` and every one after uses `daysUntilNextPurchase` :

File: src/components/List.js
Comment (veeteeran) [inline src/components/List.js:87]: This is the documentation on the [differenceInDays](https://date-fns.org/v2.17.0/docs/differenceInDays) function from date-fns

**Review by MikeBLambert — APPROVED** [review]
Great job on a really tough issue! Looks great!

**Pull request merged**
